src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,297 @@
+      CharSequence activityLabel = info.loadLabel(packageManager);
+      /** setCurrentScreen should only be called in the onResume() activity */
+      mFirebaseAnalytics.setCurrentScreen(
+              activity, activityLabel.toString(), null /* class override */);
+      logger.verbose(
+              "mFirebaseAnalytics.setCurrentScreen(%s, %s, null /* class override */);",
+              activity, activityLabel.toString());
 
+    } catch (PackageManager.NameNotFoundException e) {
+      throw new AssertionError("Activity Not Found: " + e.toString());
+    }
   }
 
-  @Override public void identify(IdentifyPayload identify) {
+  @Override
+  public void identify(IdentifyPayload identify) {
     super.identify(identify);
 
     if (!isNullOrEmpty(identify.userId())) {
       mFirebaseAnalytics.setUserId(identify.userId());
-    }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
+      Map<String, Object> traits = identify.traits();
+
+      for (Map.Entry<String, Object> entry : traits.entrySet()) {
+        String trait = entry.getKey();
+
+        trait = trait.trim().replaceAll(" ", "_");
+
+        if (trait.length() > 40) {
+          trait = trimKey(trait);
+        }
+
+        String formattedValue;
+
+        if (entry.getValue() instanceof Date) {
+
+          Date value = (Date) entry.getValue();
+          formattedValue = formatDate(value);
+
+        } else if (entry.getValue() instanceof String && ISO_DATE.matcher(String.valueOf(entry.getValue())).matches()) {
+
+          formattedValue = String.valueOf(entry.getValue()).substring(0, 10);
+
+        } else {
 
+          formattedValue = String.valueOf(entry.getValue());
+        }
+
+        mFirebaseAnalytics.setUserProperty(trait, formattedValue);
+        logger.verbose("mFirebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+      }
+    }
   }
 
-  @Override public void screen(ScreenPayload screen) {
+  @Override
+  public void screen(ScreenPayload screen) {
     super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+
+    logger.verbose(
+            "The Firebase SDK gathers screen calls automatically. Segment does map manual screen" +
+                    "methods to Firebase.");
+
   }
 
-  @Override public void track(TrackPayload track) {
+  @Override
+  public void track(TrackPayload track) {
     super.track(track);
 
+    String event = track.event();
+    String eventName = mapEvent(event);
+
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+
+    mFirebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("mFirebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
+  }
+
+  private String mapEvent(String event) {
+
+    String eventName = event;
+
+    final Map<String, String> eventMapper = new HashMap<>();
```

Comment:
This should be created once (as a `private static final` variable)  and re-used. Otherwise it will allocate and populate a map each time there is an event which is expensive.